### ISSUE TITLE
NPE in Http1xServerConnection

### ISSUE THREAD

**Issue by Brunaldoo** [eclipse-vertx/vert.x#3310]
Vert.x version 3.7.0

We are getting these rare errors (happened around 9 times in the last 90 days).

Not exactly sure how to reproduce these but it seems like this happens during high load.

```
Unhandled exception java.lang.NullPointerException at io.vertx.core.http.impl.Http1xServerConnection.handleInterestedOpsChanged(Http1xServerConnection.java:415)
	at io.vertx.core.net.impl.VertxHandler.lambda$channelWritabilityChanged$3(VertxHandler.java:136)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:188)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:180)
	at io.vertx.core.net.impl.VertxHandler.channelWritabilityChanged(VertxHandler.java:136)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:436)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:423)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.handler.stream.ChunkedWriteHandler.channelWritabilityChanged(ChunkedWriteHandler.java:150)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:436)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:423)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:416)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelWritabilityChanged(DefaultChannelPipeline.java:1431)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:436)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:423)
	at io.netty.channel.DefaultChannelPipeline.fireChannelWritabilityChanged(DefaultChannelPipeline.java:942)
	at io.netty.channel.ChannelOutboundBuffer$2.run(ChannelOutboundBuffer.java:620)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:405)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:906)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by vietj** [comment]
have you tried using a more recent version ?

**Comment by Brunaldoo** [comment]
@vietj we have recently opened a new ticket to upgrade to 3.8.5 , will get back to you after upgrading and analysing the logs to see if this happens again.

thanks.

**Comment by vietj** [comment]
yes, I will try reproduce it with 3.7.0 and see if it happens later

**Comment by vietj** [comment]
2 questions @Brunaldoo 

- does you app write to the `HttpServerResponse` from outside its event loop ?
- are you client using HTTP pipelining ?

**Comment by vietj** [comment]
I was able to reproduce this with

```java
  @Repeat(times = 1000)
  @Test
  public void testServerResponseClosePipelined() throws Exception {
    int n = 10;
    Buffer chunk = Buffer.buffer(TestUtils.randomAlphaString(8192));
    waitFor(n);
    AtomicInteger reqCnt = new AtomicInteger();
    server.requestHandler(req -> {
      HttpServerResponse resp = req.response()
        .setChunked(true)
        .setWriteQueueMaxSize(2048);
      new Thread(() -> {
        while (!resp.writeQueueFull()) {
          resp.write(chunk);
        }
        resp.end();
      }).start();
    });
    startServer();
    client.close();
    client = vertx.createHttpClient(createBaseClientOptions()
      .setMaxWaitQueueSize(n)
      .setPipelining(true)
      .setMaxPoolSize(1));
    for (int i = 0;i < n;i++) {
      client.getNow(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI, resp -> {
        resp.endHandler(v -> complete());
      });
    }
    await();
  }
```

In this reproducer, it happens when the HTTP response is written outside the request event-loop with enough buffers such that the outbound buffer is full and the client does pipelining.

There is a race condition in which the server ending the response will set the connection field `responseInProgress` and call the `handleNext` method that will start the pending pipelined HTTP request. This call ensures that the request will be execute on the connection event-loop and thus begins the request asynchronously, so there is a short amount of time where the field `responseInProgress` is set but its `response` field is still null.

When the outbound buffers empties, the connection will process the empty event and calls `handleInterestedOpsChanged` which assume that the `response` field of `responseInProgress` is never null.

```java
  @Override
  public synchronized void handleInterestedOpsChanged() {
    if (!isNotWritable()) {
      if (responseInProgress != null) {
        responseInProgress.response().handleDrained(); // Assumes it is not null
      } else if (ws != null) {
        ws.handleDrained();
      }
    }
  }
```

And that's how the NPE happens.

**Comment by vietj** [comment]
The fix is quite simple and `handleInterestedOpsChanged` should check the response is not null when calling `handleDrained()`

**Comment by vietj** [comment]
This bug does not happen in 3.8.5 release because code was rewritten in a way the bug does not produce anymore due to this bug fix https://github.com/eclipse-vertx/vert.x/issues/3082

**Comment by vietj** [comment]
I will close this issue

**Comment by vietj** [comment]
I keep this issue in 3.9.0 to signal the bug is fixed even if it was fixed before 3.9
